### PR TITLE
Allow parentPostMessage for FIE for amp-analytics

### PR DESCRIPTION
**Pull request by riklund**

### CODE REVIEW

**Comment by riklund** [comment]
@lannka @keithwrightbos @zhouyx

**Comment by zhouyx** [comment]
@dvoytenko @ampproject/wg-runtime 
Hi runtime team. This PR enables the amp-analytics component to send customized message to parent doc using postMessage. Want to make sure this won't break viewer. Please let us know. Thank you

**Comment by zhouyx** [comment]
Thought about this feature a bit more. Maybe we can restrict the usage to ads only, that are inabox and FIE.

**Comment by lannka** [comment]
> This PR enables the amp-analytics component to send customized message to parent doc using postMessage.

I think this should be fine. Viewer should not have any assumption on what messages it will receive.

**Comment by zhouyx** [comment]
> I think this should be fine. Viewer should not have any assumption on what messages it will receive.

I thought the API is based on the messages? Even not, there are a number of different viewer implementations out there, think it is still possible that sending them customized message could lead to some errors that we can't foresee.

**Comment by riklund** [comment]
Added you to an email thread with context.

**Comment by riklund** [comment]
As discussed, I've updated the pull request to restrict to the FIE case. Please take another look.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 1]
```diff
@@ -621,7 +632,7 @@ export class AmpAnalytics extends AMP.BaseElement {
    */
   expandAndPostMessage_(trigger, event) {
     const msg = trigger['parentPostMessage'];
-    if (!msg || !this.isInabox_) {
+    if (!msg || !this.allowParentPostMessage_()) {
       // Only send message in inabox runtime with parentPostMessage specified.
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/amp-analytics.js]: can we please update the comment as well. only send message for AMP ad with parentPostMessage specified. Thanks.

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 2]
```diff
@@ -1762,7 +1762,7 @@ describes.realWin(
       });
     });
 
-    describe('parentPostMessage in inabox case', () => {
+    describe('parentPostMessage', () => {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: Can we please add one more test to cover the case that `parentPostMessage` is supported in FIE. Thanks

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 3]
```diff
@@ -240,6 +240,14 @@ export class AmpAnalytics extends AMP.BaseElement {
     return this.iniPromise_;
   }
 
+  /**
+   * @return {boolean} whether parent post messages are allowed.
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Nit: maybe document our decision here. That `parentPostMessage` feature is only allowed for ads.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 4]
```diff
@@ -240,6 +240,14 @@ export class AmpAnalytics extends AMP.BaseElement {
...
+   * @private
+   */
+  allowParentPostMessage_() {
+    return this.isInabox_ || isInFie(this.element);
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/amp-analytics.js]: The `isInFie` check is expensive. Let's cache this to a property.

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js (hunk 2 above)
Comment (riklund) [reply extensions/amp-analytics/0.1/test/test-amp-analytics.js]: done.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 1 above)
Comment (riklund) [reply extensions/amp-analytics/0.1/amp-analytics.js]: done.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 3 above)
Comment (riklund) [reply extensions/amp-analytics/0.1/amp-analytics.js]: Sure. Done.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 4 above)
Comment (riklund) [reply extensions/amp-analytics/0.1/amp-analytics.js]: done.

**Pull request merged**